blockchain/pallets/minting/src/lib.rs

```diff
@@ -190,9 +190,18 @@ pub mod pallet {
             }
 
             IdDatasets::<T>::insert(id, extension_proof);
-            NextMintingRewards::<T>::insert(id, who);
 
-            Ok(())
+            // Only first call is free, determined by presence of
+            // entry in NextMintingRewards. This entry will be removed when
+            // minting rewards get dispersed in the on_finalize block. Not present
+            // indicates its the first call.
+            if NextMintingRewards::<T>::try_get(id).is_ok() {
+                // None indicates use pallet::weight annotation
+                Ok(None.into())
+            } else {
+                NextMintingRewards::<T>::insert(id, who);
```

Comment:
I think if users are already registered for minting and they want to register again for the same period, we should allow it. If only for eventually using the merkle_tree for proving data existence.